ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java

```diff
@@ -712,37 +897,106 @@ IndexValue markAsPermanent(StoreKey id, FileSpan fileSpan, long operationTimeMs)
    * @return the {@link IndexValue} of the ttl update record
    * @throws StoreException if there is any problem writing the index record
    */
-  private IndexValue markAsPermanent(StoreKey id, FileSpan fileSpan, MessageInfo info, long operationTimeMs)
-      throws StoreException {
+  private IndexValue markAsPermanent(StoreKey id, FileSpan fileSpan, MessageInfo info, long operationTimeMs,
+      short lifeVersion) throws StoreException {
     validateFileSpan(fileSpan, true);
+    boolean hasLifeVersion = IndexValue.hasLifeVersion(lifeVersion);
     IndexValue value = findKey(id);
     if (value == null && info == null) {
       throw new StoreException("Id " + id + " not present in index " + dataDir, StoreErrorCodes.ID_Not_Found);
-    } else if (value != null && value.isFlagSet(IndexValue.Flags.Delete_Index)) {
+    }
+    short retrievedLifeVersion = value == null ? info.getLifeVersion() : value.getLifeVersion();
+    if (value != null && value.isFlagSet(IndexValue.Flags.Delete_Index)) {
       throw new StoreException("Id " + id + " deleted in index " + dataDir, StoreErrorCodes.ID_Deleted);
     } else if (value != null && value.isFlagSet(IndexValue.Flags.Ttl_Update_Index)) {
       throw new StoreException("TTL of " + id + " already updated in index" + dataDir, StoreErrorCodes.Already_Updated);
+    } else if (hasLifeVersion && retrievedLifeVersion > lifeVersion) {
+      throw new StoreException("LifeVersion conflict in index. Id " + id + " LifeVersion: " + retrievedLifeVersion
+          + " Undelete LifeVersion: " + lifeVersion, StoreErrorCodes.Life_Version_Conflict);
     }
     long size = fileSpan.getEndOffset().getOffset() - fileSpan.getStartOffset().getOffset();
     IndexValue newValue;
+
     if (value == null) {
       // It is possible that the PUT has been cleaned by compaction
+      // but the TTL update is going to still be placed?
+      if (!hasLifeVersion) {
+        throw new StoreException("MessageInfo of ttlUpdate carries invalid lifeVersion",
+            StoreErrorCodes.Initialization_Error);
```

Comment:
If it doesn't have lifeVersion, it mean lifeVersion is -1. This will never happen, for 1 when the method is called in recovery or replication, we already the lifeVersion from MessageInfo 2 when the method is called in handling frontend requests, we will have a put record in the persistent index so the value == null will be false.